zerver/tests/test_push_notifications.py

```diff
@@ -1146,12 +1176,33 @@ def test_modernize_apns_payload(self) -> None:
         self.assertEqual(
             modernize_apns_payload(
                 {'alert': 'Message from Hamlet',
-                 'message_ids': [3]}),
+                 'message_ids': [3],
+                 'badge': 0}),
             payload)
         self.assertEqual(
             modernize_apns_payload(payload),
             payload)
 
+    @mock.patch('zerver.lib.push_notifications.push_notifications_enabled', return_value = True)
+    def test_apns_badge_count(self, mock_push_notifications: mock.MagicMock) -> None:
+        user_profile = self.example_user('othello')
+        # Test APNs badge count for personal messages.
+        for i in range(5):
+            self.send_personal_message(self.sender, user_profile, 'Content of message')
+        self.assertEqual(get_apns_badge_count(user_profile), 5)
+        # Mark all messages as read and test whether
+        # the count comes back to zero.
+        do_mark_all_as_read(user_profile, self.client)
+        self.assertEqual(get_apns_badge_count(user_profile), 0)
+
+        # Similarly, test APNs badge count for stream mention.
+        stream = self.subscribe(user_profile, "Denmark")
+        self.send_stream_message(self.sender, stream.name, 'Hi, @**Othello, the Moor of Venice**')
+        self.assertEqual(get_apns_badge_count(user_profile), 1)
+        do_mark_stream_messages_as_read(user_profile, self.client, stream)
+        self.assertEqual(get_apns_badge_count(user_profile), 0)
+        mock_push_notifications.assert_called()
```

Comment:
Our tests are going to need to have the badge count go above 1 in order to verify the logic correctly.